src/gui/PluginPinConnectorView.cpp

```diff
@@ -0,0 +1,457 @@
+	if (model->in().channelCount != 0)
+	{
+		const auto xwIn = std::pair{0, inMatrixRect.left() - 4};
+		int yPos = inMatrixRect.y();
+		for (unsigned idx = 0; idx < model->trackChannelsUsed(); ++idx)
+		{
+			p.drawText(xwIn.first, yPos, xwIn.second, textSize.height(), Qt::AlignRight,
+				QString::fromUtf16(u"%1 \U0001F82E").arg(idx + 1));
+
+			yPos += cellSize.height();
+		}
+	}
+
+	// Draw track channel text (out)
+	if (model->out().channelCount != 0)
+	{
+		const auto xwOut = std::pair{outMatrixRect.right() + 4, width() - outMatrixRect.right() - 4};
+		int yPos = outMatrixRect.y();
+		for (unsigned idx = 0; idx < model->trackChannelsUsed(); ++idx)
+		{
+			p.drawText(xwOut.first, yPos, xwOut.second, textSize.height(), Qt::AlignLeft,
+				QString::fromUtf16(u"\U0001F82E %1").arg(idx + 1));
+
+			yPos += cellSize.height();
+		}
+	}
+
+	p.save();
+
+	// Draw plugin channel text (in)
+	int yPos = inMatrixRect.top() - 4;
+	int xPos = inMatrixRect.x() + 2;
+	for (int idx = 0; idx < model->in().channelCount; ++idx)
+	{
+		const auto transform = QTransform{}.translate(xPos, yPos).rotate(-90);
+		p.setTransform(transform);
+
+		const auto name = model->in().channelName(idx);
+		p.drawText(0, 0, textSize.width(), yPos, Qt::AlignLeft, name);
+
+		xPos += cellSize.width();
+	}
+
+	// Draw plugin channel text (out)
+	yPos = outMatrixRect.top() - 4;
+	xPos = outMatrixRect.x() + 2;
+	for (int idx = 0; idx < model->out().channelCount; ++idx)
+	{
+		const auto transform = QTransform{}.translate(xPos, yPos).rotate(-90);
+		p.setTransform(transform);
+
+		const auto name = model->out().channelName(idx);
+		p.drawText(0, 0, textSize.width(), yPos, Qt::AlignLeft, name);
+
+		xPos += cellSize.width();
+	}
+
+	p.restore();
+}
+
+void PluginPinConnectorView::updateGeometry()
+{
+	m_inView->updateSize();
+	m_outView->updateSize();
+
+	m_subWindow->setSizePolicy(QSizePolicy::MinimumExpanding, QSizePolicy::MinimumExpanding);
+	m_subWindow->setMinimumSize(minimumSizeHint());
+
+	update();
+}
+
+//////////////////////////////////////////////
+
+PluginPinConnectorView::MatrixView::MatrixView(PluginPinConnectorView* view,
+	const PluginPinConnector::Matrix& matrix, bool isIn)
+	: QWidget{nullptr}
+	, m_model{view->castModel<PluginPinConnector>()}
+	, m_matrix{&matrix}
+{
+	assert(m_model != nullptr);
+	connect(m_model, &PluginPinConnector::dataChanged, this, static_cast<void(QWidget::*)()>(&QWidget::update));
+
+	const Model* parentModel = m_model->parentModel();
+	assert(parentModel != nullptr);
+
+	const char* formatText = isIn ? "LMMS to %1 input(s)" : "%1 output(s) to LMMS";
+	setToolTip(QString{formatText}.arg(parentModel->fullDisplayName()));
+
+	setMouseTracking(true);
+
+	updateSize();
+}
+
+auto PluginPinConnectorView::MatrixView::sizeHint() const -> QSize
+{
+	return calculateSize();
+}
+
+auto PluginPinConnectorView::MatrixView::minimumSizeHint() const -> QSize
+{
+	return calculateSize();
+}
+
+void PluginPinConnectorView::MatrixView::paintEvent(QPaintEvent*)
```

Comment:
If I made each individual cell a `QWidget` that I add to a grid layout, it might be difficult to update the view when the number of plugin channels or track channels changes. I might have to remove all the widgets from the layout and then add them all back in to get it to work.